### PR TITLE
Fix some build issues on Windows

### PR DESCRIPTION
**Pull request by grynspan**
Fix some build issues specific to the Windows implementation.

### CODE REVIEW

**Comment by grynspan** [comment]
@swift-ci please test

File: stdlib/public/CommandLineSupport/CommandLine.cpp [hunk 1]
```diff
@@ -124,25 +124,27 @@ char **_swift_stdlib_getUnsafeArgvArgc(int *outArgLen) {
     if (szBufferSize == 0) {
       swift::fatalError(0,
                         "Fatal error: Could not retrieve commandline "
-                        "arguments: %u\n",
-                        GetLastError());
+                        "argument %d: %lu\n",
+                        i, static_cast<unsigned long>(GetLastError()));
```
Comment (compnerd) [inline stdlib/public/CommandLineSupport/CommandLine.cpp:128]: Nit: The cast should be unnecessary: `DWORD GetLastError();`.

File: stdlib/public/CommandLineSupport/CommandLine.cpp [hunk 2]
```diff
@@ -124,25 +124,27 @@ char **_swift_stdlib_getUnsafeArgvArgc(int *outArgLen) {
...
 
     char *buffer = static_cast<char *>(
         calloc(static_cast<size_t>(szBufferSize), sizeof(char)));
     if (buffer == nullptr) {
       swift::fatalError(0,
-                        "Fatal error: Could not allocate space for commandline"
-                        "arguments");
+                        "Fatal error: Could not allocate space for commandline "
+                        "argument %d: %d\n",
+                        i, errno);
```
Comment (compnerd) [inline stdlib/public/CommandLineSupport/CommandLine.cpp:138]: `errno` will pretty much be `ENOMEM` no?

File: stdlib/public/CommandLineSupport/CommandLine.cpp [hunk 3]
```diff
@@ -124,25 +124,27 @@ char **_swift_stdlib_getUnsafeArgvArgc(int *outArgLen) {
...
       return nullptr;
     }
 
     if (!WideCharToMultiByte(CP_UTF8, WC_ERR_INVALID_CHARS, szArgList[i], -1,
                              buffer, szBufferSize, nullptr, nullptr)) {
       swift::fatalError(0,
                         "Fatal error: Conversion to UTF-8 failed for "
-                        "commandline arguments");
+                        "commandline argument %d: %lu\n",
+                        i, static_cast<unsigned long>(GetLastError()));
```
Comment (compnerd) [inline stdlib/public/CommandLineSupport/CommandLine.cpp:147]: Nit: Please remove the cast.

File: stdlib/public/Concurrency/Task.cpp [hunk 4]
```diff
@@ -1267,8 +1267,10 @@ static void swift_task_asyncMainDrainQueueImpl() {
 #if defined(_WIN32)
   static void(FAR *pfndispatch_main)(void) = NULL;
 
-  if (pfndispatch_main)
-    return pfndispatch_main();
+  if (pfndispatch_main) {
+    pfndispatch_main();
+    exit(0);
```
Comment (compnerd) [inline stdlib/public/Concurrency/Task.cpp:1272]: ... 🤦‍♂️ why did I not think of that?

File: stdlib/public/stubs/Random.cpp [hunk 5]
```diff
@@ -70,7 +70,8 @@ void swift_stdlib_random(void *buf, __swift_size_t nbytes) {
                                     static_cast<ULONG>(nbytes),
                                     BCRYPT_USE_SYSTEM_PREFERRED_RNG);
   if (!BCRYPT_SUCCESS(status)) {
-    fatalError(0, "Fatal error: 0x%.8X in '%s'\n", status, __func__);
+    fatalError(0, "Fatal error: 0x%lX in '%s'\n",
```
Comment (compnerd) [inline stdlib/public/stubs/Random.cpp:73]: Nit: does it make sense to use `lx` while here?

File: stdlib/public/stubs/Random.cpp [hunk 6]
```diff
@@ -70,7 +70,8 @@ void swift_stdlib_random(void *buf, __swift_size_t nbytes) {
...
+        static_cast<unsigned long>(status), __func__);
```
Comment (compnerd) [inline stdlib/public/stubs/Random.cpp:74]: Nit: please drop the cast, `NTSTATUS` is the same as `HRESULT`, which is just a `DWORD`.

**Review by compnerd — APPROVED** [review]
Thanks!

File: stdlib/public/Concurrency/Task.cpp (hunk 4 above)
Comment (compnerd) [reply stdlib/public/Concurrency/Task.cpp:1272]: Actually, please change the exit code to something `[^0-2]`.  This is a failure case.

File: stdlib/public/CommandLineSupport/CommandLine.cpp (hunk 1 above)
Comment (grynspan) [reply stdlib/public/CommandLineSupport/CommandLine.cpp:128]: It's my pedantry: I'm printing an unsigned long, so I want to explicitly pass an unsigned long where printing happens.

File: stdlib/public/CommandLineSupport/CommandLine.cpp (hunk 2 above)
Comment (grynspan) [reply stdlib/public/CommandLineSupport/CommandLine.cpp:138]: Or `EINVAL` or `ENOSYS` or `EPERM` or…

Comment (grynspan) [reply stdlib/public/CommandLineSupport/CommandLine.cpp:138]: (But yes, in practice we can expect it to always be `ENOMEM`. This is an unexpected failure either way.)

File: stdlib/public/stubs/Random.cpp (hunk 5 above)
Comment (grynspan) [reply stdlib/public/stubs/Random.cpp:73]: I have no preference here regarding uppercase or lowercase. :)

File: stdlib/public/CommandLineSupport/CommandLine.cpp (hunk 1 above)
Comment (grynspan) [reply stdlib/public/CommandLineSupport/CommandLine.cpp:128]: The pedantry can also go the other way, saying that the cast would suppress a warning if `DWORD` and `unsigned long` were _not_ the same. Will change. :)

File: stdlib/public/Concurrency/Task.cpp (hunk 4 above)
Comment (grynspan) [reply stdlib/public/Concurrency/Task.cpp:1272]: If we want to assert this function can never return, but cannot do so with `_Noreturn` (and I couldn't figure out a way to make the compiler happy with that decoration on the function pointer above), we should use `swift_unreachable()`.

File: stdlib/public/CommandLineSupport/CommandLine.cpp (hunk 2 above)
Comment (compnerd) [reply stdlib/public/CommandLineSupport/CommandLine.cpp:138]: It was more to make sure I understood what the failure modes were that we should/could expect :).  I agree with the change in principle.

**Review by compnerd — APPROVED** [review]
This looks great!

**Comment by grynspan** [comment]
@swift-ci please test

File: stdlib/public/Concurrency/Task.cpp [hunk 7]
```diff
@@ -1265,23 +1265,21 @@ static void swift_task_asyncMainDrainQueueImpl() {
...
-  pfndispatch_main =
-      reinterpret_cast<void (FAR *)(void)>(GetProcAddress(hModule,
-                                                          "dispatch_main"));
-  if (pfndispatch_main == NULL)
-    swift_reportError(0, "unable to locate dispatch_main in dispatch.dll");
+  auto pfndispatch_main = reinterpret_cast<void (FAR *)(void)>(
+    GetProcAddress(hModule, "dispatch_main"));
+  if (pfndispatch_main == NULL) {
+    swift_Concurrency_fatalError(0,
+      "unable to locate dispatch_main in dispatch.dll", GetLastError());
```
Comment (grynspan) [inline stdlib/public/Concurrency/Task.cpp:1278]: _Hoisted by my own petard!_

**Comment by grynspan** [comment]
@swift-ci please test

**Review by kubamracek — APPROVED** [review]

**Pull request merged**